DDD-1.md

```diff
@@ -0,0 +1,173 @@
+# DDD-1: Support for multiple connector tasks for relational Debezium connectors
+
+## Motivation
+
+Currently, it is challenging to use Debezium Connector for SQL Server in a multi-tenanted environment where a single SQL Server instance hosts tens to hundreds of similar tenant-scoped databases due to the fact that a given connector instance can capture changes only from a single database. The challenges are:
+
+1. Kafka Connect resources. Deploying hundreds of connector instances will require a Kafka Connect cluster with significant CPU and memory resources.
+2. Cluster stability. If the resources above are insufficient (e.g. memory), it may cause a Connect worker crash which in turn will fail all connector tasks scheduled to the worker. In the best case scenario, those tasks can be just restarted. In a worse case scenario, the restarted connectors will produce duplicates since the source connectors currently don't implement exactly-once processing semantics. In an even worse case, a worker can fail during a long-running snapshot, handling which will require additional cleanup.
+3. SQL Server resources. Each server instance will have to maintain an extra set of persistent connections (one per database, tens to hundreds in total) which may require additional server resources.
+
+## Proposed Changes
+
+### Source partition awareness at the task level ([#15](https://github.com/sugarcrm/debezium/pull/15))
+
+Currently, a Debezium connector task can process a single source partition throught its lifecycle. This partition is usually hard-coded in the context initialization code. E.g.:
+
+```java
+public SqlServerOffsetContext(SqlServerConnectorConfig connectorConfig) {
+        partition = Collections.singletonMap(SERVER_PARTITION_KEY, connectorConfig.getLogicalName());
+```
+
+In order to achieve partition-awareness, we need to introduce an API and a set of components that do the following:
+
+1. Describe a *set* of source partitions to be processed by the task.
```

Comment:
Would a source partition be a single database? Or table?

Reply:
In our case, it's a single database. Although, the API at the framework level is generic ([`TaskPartition`](https://github.com/sugarcrm/debezium/blob/DBZ-2975/debezium-core/src/main/java/io/debezium/connector/common/TaskPartition.java)), so it allows a given connector to implement any partitioning model ([`SqlServerTaskPartition`](https://github.com/sugarcrm/debezium/blob/DBZ-2975/debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerTaskPartition.java)). It should be even possible for a single connector to define different partitioning models (e.g. per-table during the snapshot).